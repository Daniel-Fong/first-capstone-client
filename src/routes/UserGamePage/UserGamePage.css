.user-game-page-div h1 {
    font-family: 'Federant', cursive;
}