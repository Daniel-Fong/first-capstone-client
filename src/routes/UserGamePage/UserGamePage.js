import React from 'react'
import TokenService from '../../services/token-service'
import PlayersList from '../../components/PlayersList/PlayersList'
import config from '../../config'
import PlayerOption from '../../components/PlayerOption/PlayerOption'

class UserGamePage extends React.Component {
    state = {
        playersForSelect: [],
        players: [],
        game: {},
        pig: [],
        error: null
    }

    componentDidMount() {
        this.fetchPlayersByGameId(this.props.match.params.game_id)
            .then(players => {
                this.setState({players: players}) 
            })
        this.fetchGameById(this.props.match.params.game_id)
            .then(game => {
              this.setState({ game: game })
            })
        this.fetchPlayersByUserId()
            .then(players => {
              console.log(players)
              this.setState({playersForSelect: players})
            })
        // this.fetchPlayersNotInGame(this.props.match.params.game_id)
        //     .then((players) => {
        //         this.setState({playersNotInGame: players})
        //     })
    }

    handleAddPlayerToGame = ( playerid ) => {
      const gameid = this.state.game.id
      const newPlayerInGame = { playerid, gameid }
      return(
        fetch(`${config.API_ENDPOINT}/players/pig`, {
          method: 'POST',
          headers: {
            'authorization': `bearer ${TokenService.getAuthToken()}`,
            'Content-type': 'application/json'
          },
          body: JSON.stringify(newPlayerInGame)
        })
       .then(res => {
        if (!res.ok)
          return res.json().then(e => Promise.reject(e))
        return res.json()
      })
      .then(playeringame => {
        this.setState({pig: [playeringame]})
      })
      .catch(error => {
        console.error({ error })
      })
    )}

    handleDeletePIG = (playerid) => {
      return(
        fetch(`http://localhost:8080/api/players/pig/${playerid}/${this.state.game.id}`, {
          method: 'DELETE',
          headers: {
            'authorization': `bearer ${TokenService.getAuthToken()}`,
            'Content-type': 'application/json'
          }
        })
          .then(res => {
            this.setState({pig: this.state.pig.filter(pig=> pig.playerid !== playerid)})
          })         
      )}

    fetchPlayersByUserId() {
      return(
          fetch(`${config.API_ENDPOINT}/players`, {
              headers: {
                  'authorization': `bearer ${TokenService.getAuthToken()}`,
              },
          })
            .then(res => {
              if (res.ok) {
                return res.json();
              }
              return Promise.reject('Error fetching players from server');
            })
            .catch(err => {
              this.setState({error: err})
            })
          )
        }

    // fetchPlayersNotInGame() {
    //   return(
    //       fetch(`${config.API_ENDPOINT}/players/list/:gameid`, {
    //           headers: {
    //               'authorization': `bearer ${TokenService.getAuthToken()}`,
    //           },
    //       })
    //         .then(res => {
    //           if (res.ok) {
    //             return res.json();
    //           }
    //           return Promise.reject('Error fetching players from server');
    //         })
    //         .catch(err => {
    //           this.setState({error: err})
    //         })
    //       )
    //     }

    fetchGameById(gameid) {
      return(
        fetch(`${config.API_ENDPOINT}/games/${gameid}`, {
          headers: {
            'authorization': `bearer ${TokenService.getAuthToken()}`,
          }
        })
        .then(res => {
          if (res.ok) {
            return res.json();
          }
          return Promise.reject('Error fetching players from server')
        })
        .catch(err => {
          this.setState({error: err})
        })
    )
  } 

    fetchPlayersByGameId(gameid) {
        return(
            fetch(`${config.API_ENDPOINT}/players/${gameid}`, {
                headers: {
                  'authorization': `bearer ${TokenService.getAuthToken()}`,
                }
            })
              .then(res => {
                if (res.ok) {
                  return res.json();
                }
                return Promise.reject('Error fetching players from server')
              })
              .catch(err => {
                this.setState({error: err})
              })
          )
        } 

    render() {
        return(
            <div className='user-game-page-div'>
              <h1>{this.state.game.name}</h1>
              <p>{this.state.game.date_modified}</p>
              <p>{this.state.game.notes}</p>
              <h2>Add Player</h2>
              <form className='add-player-to-game-form'>
                <select onChange={(e) => {
                        e.preventDefault();
                        this.handleAddPlayerToGame(e.target.value)
                        console.log(this.state.pig)}
                    }>
                          { this.state.playersForSelect.map((player) => <PlayerOption key={player.id} player={player}/> )}
                </select>
                <button className='add-player-to-game-button' typ='submit'>Add</button>
              </form>
              <PlayersList render={true} players={this.state.players} gameid={this.props.match.params.game_id} handleDeletePIG={this.handleDeletePIG} />
            </div>
        )
    }
}

export default UserGamePage;