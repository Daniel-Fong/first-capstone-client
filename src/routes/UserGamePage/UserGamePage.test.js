import React from 'react'
import ReactDOM from 'react-dom';
import UserGamePage from './UserGamePage'
import renderer from 'react-test-renderer'

describe('<UserGamePage />', () => {
    it ('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<UserGamePage match={{params: {gameid: 1}}}/>, div);
        ReactDOM.unmountComponentAtNode(div);
    })
    it ('renders the ui as expected', () => {
        const mockUserGamePage= renderer
            .create(<UserGamePage match={{params: {gameid: 1}}}/>)
            .toJSON();
        expect(mockUserGamePage).toMatchSnapshot();
    })
})