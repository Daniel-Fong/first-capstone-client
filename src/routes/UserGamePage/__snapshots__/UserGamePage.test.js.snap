// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserGamePage /> renders the ui as expected 1`] = `
<div
  className="user-game-page-div"
>
  <header>
    <h1 />
    <p />
    <p />
    <button
      className="back-button"
      onClick={[Function]}
      type="submit"
    >
      Return to Home Page
    </button>
  </header>
  <form
    className="add-player-to-game-form"
  >
    <h2>
      Add Player
    </h2>
    <select
      defaultValue=""
      onChange={[Function]}
    >
      <option
        value="none"
      >
        Select a Player to Add
      </option>
    </select>
  </form>
  <div
    className="players-list-div"
  >
    <ul
      className="players-ul"
    />
  </div>
</div>
`;
